# Predicting Remaining Useful Life
<p style="margin:30px">
    <img style="display:inline; margin-right:50px" width=50% src="https://www.featuretools.com/wp-content/uploads/2017/12/FeatureLabs-Logo-Tangerine-800.png" alt="Featuretools" />
    <img style="display:inline" width=15% src="https://upload.wikimedia.org/wikipedia/commons/e/e5/NASA_logo.svg" alt="NASA" />
</p>

The general setup for the problem is a common one: we have a single table of sensor observations over time. Now that collecting information is easier than ever, most industries have already generated *time-series* type problems by the way that they store data. As such, it is crucial to be able to handle data in this form. Thankfully, built-in functionality from [Featuretools](https://www.featuretools.com) handles time varying data well. 

We'll demonstrate an end-to-end workflow using a [Turbofan Engine Degradation Simulation Data Set](https://ti.arc.nasa.gov/tech/dash/groups/pcoe/prognostic-data-repository/#turbofan) from NASA. This notebook demonstrates a rapid way to predict the Remaining Useful Life (RUL) of an engine using an initial dataframe of time-series data. There are three sections of the notebook:
1. [Understand the Data](#Step-1:-Understanding-the-Data)
2. [Generate features](#Step-2:-DFS-and-Creating-a-Model)
3. [Make predictions with Machine Learning](#Step-3:-Using-the-Model)

*If you're running this notebook yourself, note that the Challenge Dataset will be downloaded into the data folder in this repository. If you'd prefer to download the data yourself, download and unzip the file from [https://ti.arc.nasa.gov/c/13/](https://ti.arc.nasa.gov/c/6/)*.

## Highlights
* Quickly make end-to-end workflow using time-series data
* Find interesting automatically generated features

# Step 1: Understanding the Data
Here we load in the train data and give the columns names according to the `description.txt` file. 

In [1]:
import numpy as np
import pandas as pd
import featuretools as ft
import utils

utils.download_data()

data_path = 'data/train_FD004.txt'
data = utils.load_data(data_path)
data.head()

Using previously downloaded data
Loaded data with:
61249 Recordings
249 Engines
21 Sensor Measurements
3 Operational Settings


engine_no  time_in_cycles  operational_setting_1  \
index                                                     
0              1               1                42.0049   
1              1               2                20.0020   
2              1               3                42.0038   
3              1               4                42.0000   
4              1               5                25.0063   

       operational_setting_2  operational_setting_3  sensor_measurement_1  \
index                                                                       
0                     0.8400                  100.0                445.00   
1                     0.7002                  100.0                491.19   
2                     0.8409                  100.0                445.00   
3                     0.8400                  100.0                445.00   
4                     0.6207                   60.0                462.54   

       sensor_measurement_2  sensor_measurement_3  sensor_measurement_4  \
index                                                                     
0                    549.68               1343.43               1112.93   
1                    606.07               1477.61               1237.50   
2                    548.95               1343.12               1117.05   
3                    548.70               1341.24               1118.03   
4                    536.10               1255.23               1033.59   

       sensor_measurement_5         ...          sensor_measurement_14  \
index                               ...                                  
0                      3.91         ...                        8074.83   
1                      9.35         ...                        8046.13   
2                      3.91         ...                        8066.62   
3                      3.91         ...                        8076.05   
4                      7.05         ...                        7865.80   

       sensor_measurement_15  sensor_measurement_16  sensor_measurement_17  \
index                                                                        
0                     9.3335                   0.02                    330   
1                     9.1913                   0.02                    361   
2                     9.4007                   0.02                    329   
3                     9.3369                   0.02                    328   
4                    10.8366                   0.02                    305   

       sensor_measurement_18  sensor_measurement_19  sensor_measurement_20  \
index                                                                        
0                       2212                 100.00                  10.62   
1                       2324                 100.00                  24.37   
2                       2212                 100.00                  10.48   
3                       2212                 100.00                  10.54   
4                       1915                  84.93                  14.03   

       sensor_measurement_21  index                time  
index                                                    
0                     6.3670      0 2000-01-01 00:00:00  
1                    14.6552      1 2000-01-01 00:10:00  
2                     6.4213      2 2000-01-01 00:20:00  
3                     6.4176      3 2000-01-01 00:30:00  
4                     8.6754      4 2000-01-01 00:40:00  

[5 rows x 28 columns]

## NASA Run To Failure Dataset
In this dataset we have 249 engines (`engine_no`) which are monitored over time (`time_in_cycles`). Each engine had `operational_settings` and `sensor_measurements` recorded for each cycle. The **Remaining Useful Life** (RUL) is the amount of cycles an engine has left before it needs maintenance.
What makes this dataset special is that the engines run all the way until failure, giving us precise RUL information for every engine at every point in time.

To train a model that will predict RUL, we can can simulate real predictions on by choosing a random point in the life of the engine and only using the data from before that point. We can create features with that restriction easily by using [cutoff_times](https://docs.featuretools.com/automated_feature_engineering/handling_time.html) in Featuretools.

The function `make_cutoff_times` in [utils](utils.py) does that sampling for both the `cutoff_time` and the label. You can run the next cell several times and see differing results.

In [2]:
cutoff_times = utils.make_cutoff_times(data)

cutoff_times.head()

engine_no         cutoff_time  RUL
index                                    
1              1 2000-01-01 21:00:00  194
2              2 2000-01-03 10:50:00  266
3              3 2000-01-05 21:30:00  221
4              4 2000-01-08 02:30:00  177
5              5 2000-01-10 02:10:00   84

Let's walk through a row of the `cutoff_times` dataframe. In the third row, we have engine number 3. At 3:20 on January 6, the remaining useful life of engine number 3 is 213. Having a dataframe in this format tells Featuretools that the feature vector for engine number 3 should only be calculated with data from before that point in time. 

To apply Deep Feature Synthesis we need to establish an `EntitySet` structure for our data. The key insight in this step is that we're really interested in our data as collected by `engine`. We can create an `engines` entity by normalizing by the `engine_no` column in the raw data. In the next section, we'll create a feature matrix for the `engines` entity directly rather than the base dataframe of `recordings`.

In [3]:
def make_entityset(data):
    es = ft.EntitySet('Dataset')
    es.entity_from_dataframe(dataframe=data,
                             entity_id='recordings',
                             index='index',
                             time_index='time')

    es.normalize_entity(base_entity_id='recordings', 
                        new_entity_id='engines',
                        index='engine_no')

    es.normalize_entity(base_entity_id='recordings', 
                        new_entity_id='cycles',
                        index='time_in_cycles')
    return es
es = make_entityset(data)
es

Entityset: Dataset
  Entities:
    recordings [Rows: 61249, Columns: 28]
    engines [Rows: 249, Columns: 2]
    cycles [Rows: 543, Columns: 2]
  Relationships:
    recordings.engine_no -> engines.engine_no
    recordings.time_in_cycles -> cycles.time_in_cycles

# Step 2: DFS and Creating a Model
With the work from the last section in hand, we can quickly build features using Deep Feature Synthesis (DFS). The function `ft.dfs` takes an `EntitySet` and stacks primitives like `Max`, `Min` and `Last` exhaustively across entities. Feel free to try the next step with a different primitive set to see how the results differ!

In [4]:
fm, features = ft.dfs(entityset=es, 
                      target_entity='engines',
                      agg_primitives=['last', 'max', 'min'],
                      trans_primitives=[],
                      cutoff_time=cutoff_times,
                      max_depth=3,
                      verbose=True)
fm.to_csv('simple_fm.csv')

Built 290 features
Elapsed: 05:07 | Remaining: 00:00 | Progress: 100%|██████████| Calculated: 11/11 chunks


## Machine Learning Baselines
Before we use that feature matrix to make predictions, we should check how well guessing does on this dataset. We can use a `train_test_split` from scikit-learn to split our training data once and for all. Then, we'll check the following baselines:
1. Always predict the median value of `y_train`
2. Always predict the RUL as if every engine has the median lifespan in `X_train`

We'll check those predictions by finding the mean of the absolute value of the errors.

In [5]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

fm = pd.read_csv('simple_fm.csv', index_col='engine_no')
X = fm.copy().fillna(0)
y = X.pop('RUL')

X_train, X_test, y_train, y_test = train_test_split(X, y)

medianpredict1 = [np.median(y_train) for _ in y_test]
print('Baseline by median label: Mean Abs Error = {:.2f}'.format(
    mean_absolute_error(medianpredict1, y_test)))

Baseline by median label: Mean Abs Error = 63.03


In [6]:
recordings_from_train = es['recordings'].df[es['recordings'].df['engine_no'].isin(y_train.index)]
median_life = np.median(recordings_from_train.groupby(['engine_no']).apply(lambda df: df.shape[0]))

recordings_from_test = es['recordings'].df[es['recordings'].df['engine_no'].isin(y_test.index)]
life_in_test = recordings_from_test.groupby(['engine_no']).apply(lambda df: df.shape[0])-y_test

medianpredict2 = (median_life - life_in_test).apply(lambda row: max(row, 0))
print('Baseline by median life: Mean Abs Error = {:.2f}'.format(
    mean_absolute_error(medianpredict2, y_test)))

Baseline by median life: Mean Abs Error = 77.63


# Step 3: Using the Model
Now, we can use our created features to fit a `RandomForestRegressor` to our data and see if we can improve on the previous scores.

In [7]:
reg = RandomForestRegressor()
reg.fit(X_train, y_train)
    
preds = reg.predict(X_test)
scores = mean_absolute_error(preds, y_test)
print('Mean Abs Error: {:.2f}'.format(scores))
high_imp_feats = utils.feature_importances(X, reg, feats=10)


Mean Abs Error: 37.54
1: MAX(recordings.sensor_measurement_13) [0.186]
2: MAX(recordings.cycles.LAST(recordings.sensor_measurement_4)) [0.154]
3: MAX(recordings.sensor_measurement_4) [0.137]
4: MAX(recordings.cycles.LAST(recordings.sensor_measurement_11)) [0.052]
5: MIN(recordings.sensor_measurement_21) [0.033]
6: LAST(recordings.cycles.MIN(recordings.sensor_measurement_7)) [0.021]
7: MAX(recordings.sensor_measurement_3) [0.020]
8: MAX(recordings.cycles.LAST(recordings.sensor_measurement_13)) [0.017]
9: MAX(recordings.sensor_measurement_11) [0.016]
10: MAX(recordings.cycles.MAX(recordings.sensor_measurement_8)) [0.014]
-----



Next, we can apply the exact same transformations (including DFS) to our test data. For this particular case, the real answer isn't in the data so we don't need to worry about cutoff times.

In [8]:
data2 = utils.load_data('data/test_FD004.txt')
es2 = make_entityset(data2)
fm2 = ft.calculate_feature_matrix(entityset=es2, features=features, verbose=True)
fm2.head()

Loaded data with:
41214 Recordings
248 Engines
21 Sensor Measurements
3 Operational Settings
Elapsed: 00:30 | Remaining: 00:00 | Progress: 100%|██████████| Calculated: 11/11 chunks


LAST(recordings.time_in_cycles)  \
engine_no                                    
1                                      230   
2                                      153   
3                                      141   
4                                      208   
5                                       51   

           LAST(recordings.operational_setting_1)  \
engine_no                                           
1                                         25.0070   
2                                         41.9989   
3                                         42.0005   
4                                         25.0018   
5                                         25.0039   

           LAST(recordings.operational_setting_2)  \
engine_no                                           
1                                          0.6214   
2                                          0.8400   
3                                          0.8401   
4                                          0.6207   
5                                          0.6200   

           LAST(recordings.operational_setting_3)  \
engine_no                                           
1                                            60.0   
2                                           100.0   
3                                           100.0   
4                                            60.0   
5                                            60.0   

           LAST(recordings.sensor_measurement_1)  \
engine_no                                          
1                                         462.54   
2                                         445.00   
3                                         445.00   
4                                         462.54   
5                                         462.54   

           LAST(recordings.sensor_measurement_2)  \
engine_no                                          
1                                         537.66   
2                                         549.96   
3                                         549.47   
4                                         536.06   
5                                         537.36   

           LAST(recordings.sensor_measurement_3)  \
engine_no                                          
1                                        1264.31   
2                                        1354.05   
3                                        1341.06   
4                                        1253.49   
5                                        1263.60   

           LAST(recordings.sensor_measurement_4)  \
engine_no                                          
1                                        1046.41   
2                                        1133.55   
3                                        1118.90   
4                                        1038.53   
5                                        1052.52   

           LAST(recordings.sensor_measurement_5)  \
engine_no                                          
1                                           7.05   
2                                           3.91   
3                                           3.91   
4                                           7.05   
5                                           7.05   

           LAST(recordings.sensor_measurement_6)  \
engine_no                                          
1                                           8.99   
2                                           5.72   
3                                           5.69   
4                                           9.00   
5                                           9.03   

                                       ...                               \
engine_no                              ...                                
1                                      ...                                
2                                      ...                                
3                                      ...                                
4      

In [9]:
X = fm2.copy().fillna(0)
y = pd.read_csv('data/RUL_FD004.txt', sep=' ', header=-1, names=['RUL'], index_col=False)
preds2 = reg.predict(X)
print('Mean Abs Error: {:.2f}'.format(mean_absolute_error(preds2, y)))

medianpredict1 = [np.median(y_train) for _ in preds2]
print('Baseline by median label: Mean Abs Error = {:.2f}'.format(
    mean_absolute_error(medianpredict1, y)))

medianpredict2 = (median_life - es2['recordings'].df.groupby(['engine_no']).apply(lambda df: df.shape[0])).apply(lambda row: max(row, 0))
print('Baseline by median life: Mean Abs Error = {:.2f}'.format(
    mean_absolute_error(medianpredict2, y)))

Mean Abs Error: 40.33
Baseline by median label: Mean Abs Error = 52.08
Baseline by median life: Mean Abs Error = 49.55


This is the simple version of a more advanced notebook that can be found in the [second](Advanced%20Featuretools%20RUL.ipynb) notebook. That notebook will show how to use a novel entityset structure, custom primitives, and automated hyperparameter tuning to improve the score.